### PR TITLE
Fix: move `waitFor<EngineConnectedEvent>` after creating remote participants

### PR DESCRIPTION
**Pull request by hiroshihorie**
This fixes Issue https://github.com/livekit/client-sdk-flutter/issues/57

The main cause of this issue #57
was waitFor was before creating remote participants from join response.
Which probably happened in the “Create room before connect” PR.

### CODE REVIEW

File: lib/src/room.dart [hunk 1]
```diff
@@ -125,13 +125,20 @@ class Room extends DisposableChangeNotifier with EventsEmittable<RoomEvent> {
...
     for (final info in joinResponse.otherParticipants) {
       logger.fine(
           'DEBUG_01 Existing Participant: ${info.sid}(${info.identity}) tracks:${info.tracks.map((e) => e.sid)}');
-      _getOrCreateRemoteParticipant(info.sid, info);
+      final isNew = !_participants.containsKey(info.sid);
+      final participant = _getOrCreateRemoteParticipant(info.sid, info);
+      if (!isNew) {
+        participant.hasInfo;
+        // update participant in cases where track arrived first and participant was already created.
+        await participant.updateFromInfo(info);
```
Comment (hiroshihorie) [inline lib/src/room.dart:138]: we need to update it here

File: lib/src/room.dart [hunk 2]
```diff
@@ -125,13 +125,20 @@ class Room extends DisposableChangeNotifier with EventsEmittable<RoomEvent> {
...
+    // logger.fine('DEBUG_01 Did connect');
 
     for (final info in joinResponse.otherParticipants) {
       logger.fine(
           'DEBUG_01 Existing Participant: ${info.sid}(${info.identity}) tracks:${info.tracks.map((e) => e.sid)}');
-      _getOrCreateRemoteParticipant(info.sid, info);
+      final isNew = !_participants.containsKey(info.sid);
+      final participant = _getOrCreateRemoteParticipant(info.sid, info);
+      if (!isNew) {
+        participant.hasInfo;
```
Comment (davidliu) [inline lib/src/room.dart:136]: does this line do anything?

**Review by davidliu — COMMENTED** [review]
Still a little unclear here; is the track publish event arriving from the server before the new participant join event arrives?

**Comment by davidzhao** [comment]
I just want to note that for the metadata to arrive *after* onTrack could happen, but is very unlikely. Does it explain why Android seems to be the only one having issues receiving remote video tracks? (but audio would work).

**Comment by hiroshihorie** [comment]
The main issue is the waiting for **ice connect logic** was _before_ creating remote participants. Maybe in this PR https://github.com/livekit/client-sdk-flutter/commit/0f5a758cd4ff5e2e8bf12c4a5536afe058d37c1b

Now the order should be correct in this commit https://github.com/livekit/client-sdk-flutter/pull/58/commits/af9957dea304da0f63467dba81ad9bacec9dc9e4

File: lib/src/room.dart (hunk 2 above)
Comment (hiroshihorie) [reply lib/src/room.dart:126]: oops! thats not needed 😂

File: lib/src/room.dart [hunk 3]
```diff
@@ -112,6 +112,22 @@ class Room extends DisposableChangeNotifier with EventsEmittable<RoomEvent> {
...
+    );
+
+    for (final info in joinResponse.otherParticipants) {
+      logger.fine('DEBUG_01 participant: ${info.sid}(${info.identity}) '
+          'tracks:${info.tracks.map((e) => e.sid)}');
+      final isNew = !_participants.containsKey(info.sid);
+      final participant = _getOrCreateRemoteParticipant(info.sid, info);
+      if (!isNew) {
+        // update participant in cases where track arrived first and participant was already created.
+        await participant.updateFromInfo(info);
```
Comment (hiroshihorie) [inline lib/src/room.dart:127]: I am pretty sure it's very unlikely but I wonder if it's 0 possibility that between
**after receiving the join response** and **creating existing remote participants**
tracks could be received(EngineTrackAddedEvent).

I guess this code is not required and can be reverted.

**Comment by hiroshihorie** [comment]
@davidzhao @davidliu 
The main cause of this issue https://github.com/livekit/client-sdk-flutter/issues/57
was waitFor<IceConnect> was before creating remote participants from join response.
Which probably happened in the “Create room before connect” PR.

Reverted additional checks since it is very unlikely to be required.
Sorry for the confusion 😅

File: lib/src/room.dart [hunk 4]
```diff
@@ -112,24 +112,14 @@ class Room extends DisposableChangeNotifier with EventsEmittable<RoomEvent> {
...
       room: this,
       info: joinResponse.participant,
     );
 
-    // logger.fine('DEBUG_01 Did connect');
-
     for (final info in joinResponse.otherParticipants) {
-      logger.fine(
-          'DEBUG_01 Existing Participant: ${info.sid}(${info.identity}) tracks:${info.tracks.map((e) => e.sid)}');
+      logger.fine('DEBUG_01 participant: ${info.sid}(${info.identity}) '
```
Comment (davidzhao) [inline lib/src/room.dart:121]: note on DEBUG_01 logging

**Review by davidzhao — APPROVED** [review]

**Pull request merged**